sympy/ntheory/factor_.py

```diff
@@ -523,8 +523,22 @@ def perfect_power(n, candidates=None, big=True, factor=True):
         pp = perfect_power(-n)
         if pp:
             b, e = pp
-            if e % 2:
-                return -b, e
+            smallest_odd_factor = None
+            e_temp = e
+            i = 3
+            while i <= e_temp:
+                if e_temp % i == 0 and i % 2 == 1:
+                    smallest_odd_factor = i
+                    break
+                if i == 3:
+                    i = 5
+                else:
+                    i += 2
+
+            if smallest_odd_factor:
+                power = e // smallest_odd_factor
+                new_base = -int(b ** power)
+                return new_base, smallest_odd_factor
```

Comment:
I think that this should be the largest odd factor rather than the smallest:
```python
In [4]: perfect_power(5**15)
Out[4]: (5, 15)

In [5]: perfect_power((-5)**15)
Out[5]: (-3125, 3)
```
Basically just extract all factors of 2.

Then again there is the big parameter:
```python
In [6]: perfect_power(5**15, big=True)
Out[6]: (5, 15)

In [7]: perfect_power(5**15, big=False)
Out[7]: (3125, 3)
```
So maybe the smallest odd factor is wanted when `big=False` and the largest when `big=True`.

Possibly the `big` parameter should be passed to the recursive call to `perfect_power`. Not sure what to do with `candidates`.

Reply:
```python
n = as_int(n)
    if n < 0:
        # Passing the big parameter to recursive call now
        pp = perfect_power(-n, candidates=candidates, big=big, factor=factor)
        if pp:
            b, e = pp
            # Finding all odd divisors of e 
            odd_divisors = []
            for i in range(3, e + 1, 2):
                if e % i == 0 and i > 1:
                    odd_divisors.append(i)

            if odd_divisors:
                # Choosing divisor based on big parameter as mentioned in the prev comment
                chosen_factor = max(odd_divisors) if big else min(odd_divisors)
                power = e // chosen_factor
                new_base = -int(b ** power)
                return new_base, chosen_factor
        return False
```

I believe this should work fine.

Reply:
As pointed out in the SO answer [here](https://stackoverflow.com/a/79172655/1089161), `e&-e` is a fast way to find the power of 2 that can be removed from `e` to leave behind the odd exponent. What about:
```python
if n<0:
    pp = perfect_power(-n,candidates,big=True,factor=factor)
    if pp:
        b, e=pp
        e2 = e&(-e)
        e //= e2
        b = b**e2
        if not big:
            small = min(factorint(e))
            e = small
            b = b**(e//small)
        return -b, e
```

Reply:
Yes this is more efficient for sure. 
Bit manipulation instead of a loop, not having to find all the odd divisors, I guess clearer  separation between removing power of 2 and handling `big=False` case. Furthermore lesser lines of code. 

May I use this in the PR ? 

Maybe passing `big=True` should be reconsidered as it causes unexpected behavior by overriding user's choice of `big` in case `big=False` is passed in the root call, 

```python
>>> perfect_power((-5)**15, big=False)
(-5, 3) # This is incorrect
```
It calls `perfect_power(5**15, big=True)` (ignoring our `big=False`). Gets `(5, 15)` as result. Processes this result with our `big=False` logic. Returns `(-5, 3)` instead of expected `(-3125, 3)`

Maybe `big=big` being passed in the recursive call makes more sense. Correct me if I'm wrong here.

Reply:
Feel free to use. The reason to override big on recursive call is to get the big exponent and then pare it down if necessary. Since the number passed is positive, you won't enter the negative-logic block. I had two lines reversed, however: `e = small` should come after calculating the new `b`. I am correcting that now. The logic is basically:
```python
def perfect_power(n, candidates=None, big=True, factor=True):
  if n<0:
    pp = _pp(-n,candidates,big=True,factor=factor)
    if pp:
        b, e=pp
        e2 = e&(-e)
        e //= e2
        b = b**e2
        if not big:
            small = min(factorint(e))
            b = b**(e//small)
            e = small
        return -b, e
  return _pp(n,candidates,big,factor)
```

Reply:
The other thing to do is calculate small as:
```python
for small in primerange(3, e+1):
    if not e%small:
        break
```